okhttp/third_party/okhttp/java/io/grpc/okhttp/internal/Platform.java

```diff
@@ -199,14 +176,44 @@ private static Platform findPlatform() {
       Method getMethod = negoClass.getMethod("get", SSLSocket.class);
       Method removeMethod = negoClass.getMethod("remove", SSLSocket.class);
       return new JdkWithJettyBootPlatform(
-          putMethod, getMethod, removeMethod, clientProviderClass, serverProviderClass, sslProvider);
+          putMethod, getMethod, removeMethod, clientProviderClass, serverProviderClass,
+          sslProvider);
     } catch (ClassNotFoundException ignored) {
     } catch (NoSuchMethodException ignored) {
     }
 
     return new Platform(sslProvider);
   }
 
+  /**
+   * Forcibly load the conscrypt security provider on AppEngine if it's available. If not fail.
+   */
+  private static Provider getAppEngineProvider() {
+    try {
+      // Forcibly load conscrypt as it is unlikely to be an installed provider on AppEngine
+      return (Provider) Class.forName("org.conscrypt.OpenSSLProvider").newInstance();
+    } catch (Throwable t) {
+      throw new RuntimeException("Unable to load conscrypt security provider", t);
+    }
+  }
+
+  /**
+   * Select from the available security providers in preference order. If a preferred provider
+   * is not found then warn but continue.
+   */
+  private static Provider getAndroidSecurityProvider() {
+    for (String providerClassName : ANDROID_SECURITY_PROVIDERS) {
+      for (Provider availableProvider : Security.getProviders()) {
+        if (providerClassName.equals(availableProvider.getClass().getName())) {
+          logger.log(Level.INFO, "Found registered provider " + providerClassName);
```

Comment:
> Was using string concatenation instead of message formatting on purpose?

log(FINE, "Found registered provider {0}", providerClassName)

Or did you do the concatenation on purpose?